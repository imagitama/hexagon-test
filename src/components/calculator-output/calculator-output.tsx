import React from 'react'
import { connect } from 'react-redux'
import { AppState } from '../../store'
import getResultForInputText, { OutOfBoundsError, InvalidInputError } from './formatter'

interface CalculatorInputProps {
  inputText: string
}

export const CalculatorOutput = ({ inputText }: CalculatorInputProps) => {
  if (!inputText) {
    return (
      <>
        Enter a number to calculate
      </>
    )
  }

  let result = ''

  try {
    result = getResultForInputText(inputText)
  } catch (err) {
    if (err instanceof OutOfBoundsError) {
      return (
        <>
          Cannot calculate that input (over 99999.99 or less than 0)
        </>
      )
    }

    if (err instanceof InvalidInputError) {
      return (
        <>
          Invalid input
        </>
      )
    }

    return (
      <>
        Internal error
      </>
    )
  }

  return (
    <>
      The result: {result}
    </>
  )
}

export const mapStateToProps = ({ calculator: { inputText } }: AppState) => ({ inputText })

export default connect(mapStateToProps)(CalculatorOutput)