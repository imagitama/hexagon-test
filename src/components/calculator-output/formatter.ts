const numerals = {
  million: 'million',
  thousand: 'thousand',
  hundred: 'hundred'
}

const AND = 'AND'
const DOLLARS = 'DOLLARS'
const CENTS = 'CENTS'

const tensNames = ['ten', 'twenty', 'thirty', 'forty', 'fifty', 'sixty', 'seventy', 'eighty', 'ninety']

const teenNames = ['ten', 'eleven', 'twelve', 'thirteen', 'fourteen', 'fifteen', 'sixteen', 'seventeen', 'eighteen', 'nineteen']

const digitNames = ['one', 'two', 'three', 'four', 'five', 'six', 'seven', 'eight', 'nine']

export default (inputText: string) => {
  const inputRounded = parseFloat(inputText).toFixed(2)

  const chunks = inputRounded.toString().split('.')
  const dollarChunks = chunks[0].split('')
  const centChunks = chunks[1].split('')

  let result: string[] = []

  const parseHundreds = (digits: string[]) => {
    let result: string[] = []

    digits.forEach((digit, idx) => {
      if (idx === 0) {
        result.push(digitNames[parseInt(digit) - 1])
        result.push(numerals.hundred)
        return
      }

      if (idx === 1) {
        result.push(AND)
        result.push(tensNames[parseInt(digit) - 1])
        return
      }

      if (idx === 2) {
        result.push(digitNames[parseInt(digit) - 1])
        result.push(DOLLARS)
        return
      }
    })

    return result
  }

  dollarChunks.forEach((digit, idx) => {
    // single thousands
    if (dollarChunks.length === 4) {
      if (idx === 0) {
        result.push(digitNames[parseInt(digit) - 1])
        result.push(numerals.thousand)
        return
      } else if (idx === 1) {
        result.push(AND)
        result = result.concat(parseHundreds(dollarChunks.slice(1)))
      }
    }

    // double thousands
    if (dollarChunks.length === 5) {
      if (idx === 0) {
        result.push(tensNames[parseInt(digit) - 1])
        return
      } else if (idx === 1) {
        result.push(digitNames[parseInt(digit) - 1])
        result.push(numerals.thousand)
        return
      } else if (idx === 2) {
        result.push(AND)
        result = result.concat(parseHundreds(dollarChunks.slice(2)))
      }
    }

    if (dollarChunks.length === 1) {
      result.push(digitNames[parseInt(digit) - 1])
      result.push(DOLLARS)
      return
    }

    if (idx === 0 && dollarChunks.length === 2) {
      result.push(teenNames[parseInt(digit) - 1])
      result.push(DOLLARS)
      return
    }

    if (idx === 0 && dollarChunks.length === 3) {
      result = result.concat(parseHundreds(dollarChunks))
    }
  })

  centChunks.forEach((digit, idx) => {
    if (centChunks[0] === '0' && centChunks[1] === '0') {
      return
    }

    if (idx === 0) {
      result.push(AND)

      if (digit !== '0') {
        result.push(tensNames[parseInt(digit) - 1])
      }

      if (centChunks[1] === '0') {
        result.push(CENTS)
      }

      return
    }

    if (idx === 1 && digit !== '0') {
      result.push(digitNames[parseInt(digit) - 1])
      result.push(CENTS)
      return
    }
  })



  return result.join(' ').toUpperCase()
}