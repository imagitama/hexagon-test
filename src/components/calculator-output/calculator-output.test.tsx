import React from 'react'
import { shallow, ShallowWrapper } from 'enzyme'
import { CalculatorOutput, mapStateToProps } from './calculator-output'
import getResultForInputText from './formatter'

jest.mock('./formatter')

describe('Calculator output', () => {
  describe('When mounted', () => {
    let component: ShallowWrapper

    beforeAll(() => {
      component = shallow(<CalculatorOutput inputText="" />)
    })

    describe('With no input text', () => {
      beforeAll(() => {
        component.setProps({
          inputText: ''
        })
      })

      it('Renders correctly', () => {
        expect(component).toMatchSnapshot()
      })
    })

    describe('With invalid input - not a number', () => {
      beforeAll(() => {
        component.setProps({
          inputText: 'Hello world!'
        })
      })

      it('Renders correctly', () => {
        expect(component).toMatchSnapshot()
      })
    })

    describe('With invalid input - over the limit', () => {
      beforeAll(() => {
        component.setProps({
          inputText: '9999999999999999999999'
        })
      })

      it('Renders correctly', () => {
        expect(component).toMatchSnapshot()
      })
    })

    describe('When the formatter fails', () => {
      beforeAll(() => {
        getResultForInputText.mockImplementation(() => {
          throw new Error('Something bad happened')
        })

        component.setProps({
          inputText: '123.456'
        })
      })

      it('Renders correctly', () => {
        expect(component).toMatchSnapshot()
      })
    })

    describe('When the formatter succeeds', () => {
      beforeAll(() => {
        getResultForInputText.mockImplementation(jest.requireActual('./formatter').getResultForInputText)

        component.setProps({
          inputText: '123.456'
        })
      })

      it('Renders correctly', () => {
        expect(component).toMatchSnapshot()
      })
    })
  })

  describe('mapStateToProps', () => {
    describe('When called with mock state', () => {
      it('Maps to props correctly', () => {
        expect(mapStateToProps({
          calculator: {
            inputText: 'Hello world!'
          }
        })).toEqual({
          inputText: 'Hello world!'
        })
      })
    })
  })
})
