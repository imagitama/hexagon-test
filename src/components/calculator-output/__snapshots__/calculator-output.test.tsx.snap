// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator output When mounted When the formatter fails Renders correctly 1`] = `
<Fragment>
  Internal error
</Fragment>
`;

exports[`Calculator output When mounted When the formatter succeeds Renders correctly 1`] = `
<Fragment>
  The result: 
</Fragment>
`;

exports[`Calculator output When mounted With invalid input - not a number Renders correctly 1`] = `
<Fragment>
  Invalid input
</Fragment>
`;

exports[`Calculator output When mounted With invalid input - over the limit Renders correctly 1`] = `
<Fragment>
  Cannot calculate that input (over 99999.99)
</Fragment>
`;

exports[`Calculator output When mounted With no input text Renders correctly 1`] = `
<Fragment>
  Enter a number to calculate
</Fragment>
`;
